core/src/banking_stage/consumer.rs

```diff
@@ -64,19 +64,37 @@ pub struct ExecuteAndCommitTransactionsOutput {
     error_counters: TransactionErrorMetrics,
 }
 
-pub struct Consumer;
+pub struct Consumer {
+    committer: Committer,
+    transaction_recorder: TransactionRecorder,
+    qos_service: QosService,
+    log_messages_bytes_limit: Option<usize>,
+    test_fn: Option<Box<dyn Fn() + Send>>,
+}
 
 impl Consumer {
+    pub fn new(
+        committer: Committer,
+        transaction_recorder: TransactionRecorder,
+        qos_service: QosService,
+        log_messages_bytes_limit: Option<usize>,
+        test_fn: Option<Box<dyn Fn() + Send>>,
```

Comment:
Not a huge fan of having a `dyn` here, but this is only used for tests.

Reply:
how about this?

```rust
pub struct Consumer<'a, F: Fn()> {
    committer: Committer,
    transaction_recorder: TransactionRecorder,
    qos_service: QosService,
    log_messages_bytes_limit: Option<usize>,
    #[cfg(test)]
    test_fn: Option<&'a F>,
}
```

1. move to under cfg(test)
2. take a ref and make them generic over lifetime and actual closure.

Reply:
🤔 the biggest objection I have to something like this is that we then need to construct the `Consumer` differently for tests vs operation. It'd also just make the code slightly uglier since we'd be injecting a `#[cfg(test)]` into the normal code where we call the `test_fn`.

Reply:
how about this? i scrapped the generic idea. i hope this is less uglier than you had in mind.: f8d6d95d9f3791bf7c0d5aba492e97b120131efc

imo, `#[cfg(test)]` in production code is acceptable because it's protected by rustc in this case. Actually, it simplifies code read for humans. on the other hand, `#[cfg(test)]`-ing self-standing bunch of logic can be frowned upon as this is introducing some risk of hard-to-spot bug in production code.

first of all, the existence of `test_fn` is itself a bit code smell. it's just illustration that code is organized in unit-test unfriendly way. this should be avoided at all costs. but fixing this is certainly out of scope for this pr.

Reply:
> first of all, the existence of test_fn is itself a bit code smell. it's just illustration that code is organized in unit-test unfriendly way. this should be avoided at all costs. but fixing this is certainly out of scope for this pr.

100% - I'd prefer updating unit-tests so that we can completely get rid of it. I think that commit (f8d6d95) is not too bad, but I really dislike the fact that we'd have different numbers of args to the function based on whether or not we're in a test - that just feels quite wrong.

I'll take a look to see how it's used in the unit-tests...maybe it won't be too bad to just rip this out 🤔

Reply:
It's only used in a single place 🤦

I created #30615 and #30616